docs/infrastructure/vs-deps-branches.md

```diff
@@ -0,0 +1,132 @@
+   like us to consume it as soon as they have merged their feature work into the (internal) Visual Studio master branch; this means they can get API feedback and shake out any bugs as soon as possible.
+2. We want public contributors to be able to contribute to all of Roslyn, the Visual Studio layers included. Even if the contributor isn't _changing_ the code in the Visual Studio layer, deploying to Visual
+   Studio is the only way to actually _test_ any changes made to any of our layers in the middle in an interactive way. A contributor making a change to a code fix or refactoring can of course write unit
+   tests, but they can't just "try it locally" to experiment.
+
+To find a balance between these two tensions, we have two branches. Using our master branch as an example, we have two branches named 'master' and 'master-vs-deps':
+
+1. The master branch as a rule must be buildable and runnable against the latest public non-preview build of Visual Studio. This means no change can be made inside of it that is dependent on an preview
+   or yet unshipped API in a way that would break the ability to run locally.
+2. The master-vs-deps branch only has to run against whatever internal Visual Studio branch we are targeting for insertion. In some cases, if we have a breaking change that must be coordinated with a
+   partner team, this means for a small window in time it might be runnable against _no_ build because we are doing a simultaneous insertion. This latter case is rare, but does happen.
+
+We have a bot that regularly merges code from master to master-vs-deps. This means that whether a change goes into master or master-vs-deps, it'll ship in the same Visual Studio release. The distinction
+is simply what the code can be locally ran against.
+
+As a concrete example, imagine that the internal builds of Visual Studio are building 16.5. Changes in Roslyn master will ship in 16.5; so will changes in master-vs-deps. However, changes in master must
+be runnable by an external contributor, so it must be runnable on 16.4. Changes in master-vs-deps don't have to be runnable anywhere, but obviously cannot depend on something shipping in 16.6, because
+the branch is still shipping in 16.5.
+
+master is not the only branch that has -vs-deps branches. We maintain the pair for all our milestone branches. Consider the case where we are beginning to lock down a release. At this time we often
+snap master and only approved changes go in. We still have a -vs-deps branch and a regular branch, if for no other reason than we must auto-merge any further changes into the appropriate branches. Why
+not branch _just_ the -vs-deps branch? Because then all changes would go into there, and a compiler change which has no reason to be anywhere in -vs-deps would end up getting merged into master-vs-deps.
```

Comment:
This doesn't sound quite right. AFAIK, the reason is that `16.X-previewY` has not been shipped, so we cannot merge `16.X-previewY-vs-deps` into the non-vs-deps branch if we want it publicly buildable. We can't merge everything into the -vs-deps branch because then the merges will only put changes into the `16.X-previewY+1-vs-deps` branch, instead of the non-VS-deps branch it belongs in.